### ISSUE TITLE
Auto-save in new Item Submission form breaks the form

### ISSUE THREAD

**Issue by MarisArins** [DSpace/dspace-angular#835]
1. Start creating new Item in collection "Books"
2. Enter Author manually : "John Doe"
3. Enter Title "QWERTY123"
4. Switch to next field - auto save happens as Title is marked as save-on-change field
5. Author values get cleared out !! 
	BUG
6. Fill in all the mandatory fields in all the sections of create form and lastly fill the author fields again
7. The first section "Describe" is still marked as not complete/with errors
	BUG
8. Deposit button is enabled and can be pressed - press that
9. pop-up message appears about error in Deposit process
10. error response in Network tab says : 

`{"timestamp":"2020-08-13T07:11:41.940+0000","status":422,"error":"Unprocessable Entity","message":"Unprocessable or invalid entity","trace":"org.dspace.app.rest.exception.UnprocessableEntityException: Start workflow failed due to validation error on workspaceitem
	at org.dspace.app.rest.submit.SubmissionService.createWorkflowItem(SubmissionService.java:243)
	at org.dspace.app.rest.repository.WorkflowItemRestRepository.createAndReturn(WorkflowItemRestRepository.java:173)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:774)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691)
	at org.dspace.app.rest.repository.WorkflowItemRestRepository$$EnhancerBySpringCGLIB$$5dbf8ba7.createAndReturn(<generated>)
	at org.dspace.app.rest.repository.DSpaceRestRepository.createAndReturn(DSpaceRestRepository.java:306)
	at org.dspace.app.rest.repository.DSpaceRestRepository$$FastClassBySpringCGLIB$$615525f9.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:687)
	at org.dspace.app.rest.repository.WorkflowItemRestRepository$$EnhancerBySpringCGLIB$$5dbf8ba7.createAndReturn(<generated>)
	at org.dspace.app.rest.RestResourceController.postUriListInternal(RestResourceController.java:460)
	at org.dspace.app.rest.RestResourceController.postWithUriListContentType(RestResourceController.java:405)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:105)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.dspace.app.rest.filter.DSpaceRequestContextFilter.doFilter(DSpaceRequestContextFilter.java:45)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.dspace.app.rest.filter.ContentLanguageHeaderResponseFilter.doFilter(ContentLanguageHeaderResponseFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.dspace.utils.servlet.DSpaceWebappServletFilter.doFilter(DSpaceWebappServletFilter.java:84)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:118)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.dspace.app.rest.security.AnonymousAdditionalAuthorizationFilter.doFilterInternal(AnonymousAdditionalAuthorizationFilter.java:66)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.dspace.app.rest.security.StatelessAuthenticationFilter.doFilterInternal(StatelessAuthenticationFilter.java:101)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:92)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:128)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:103)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:121)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.valves.CrawlerSessionManagerValve.invoke(CrawlerSessionManagerValve.java:235)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:688)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.ajp.AjpProcessor.service(AjpProcessor.java:415)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1639)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)`

**Comment by artlowel** [comment]
Other issues with auto-save as mentioned in the meeting on Oct. 1st:

- If you're halfway through typing something when an auto-save occurs, the field you were typing in will lose focus (meaning you have to click it again to resume typing), and what you were typing was gone. You can replicate this on demo by typing something in a non-repeatable, clicking another non-repeatable filed (or using the tab key to quickly switch to it) and immediately starting to type again. The reason for this is that when the response comes back from the server after a PATCH, the entire submission form is re-rendered using data from the server. So the input instance you were typing in doesn't exist anymore after that re-render.  
To solve this we also shouldn't simply save the half-typed value as metadata for that field, because for an authority controlled or entity field, what you're typing might be a query to search for an authority/entity or it might be a plain metadata value, we can't know (as noted in [my comment on #864](https://github.com/DSpace/dspace-angular/issues/864#issuecomment-690083232)). We don't want people to have to manually remove queries from the list of metadata values because of auto-save.
- When an auto-save occurs all required fields will get a red border, even if the user hasn't interacted with them yet. The standard way of dealing with this is to add a "touched" attribute or class to a field when the user first interacts with it, and only show errors for fields with that attribute/class.
- When an auto-save occurs you get the notification "save successful" every time, and usually: “there are incomplete sections". This is too obtrusive in my opinion. Unless the user manually saves, a small little span saying 'saved n seconds ago' somewhere close to the save button would be better, if we even need that. The github editor in which I'm typing this for example, auto-saves without any notification.

To solve these issues I think we should:
- Stop autosaving on a timer, instead auto-save whenever a field loses focus
- Don't re-render the form when a server response comes in if _any_ field is currently in focus, only when a field loses focus. 
    - this could still mean we lose the tabindex, meaning if you switch fields by using the tab button (something required for a lot accessibility use cases) you could lose your place when the autosave happens
    - To counter that we could debounce the re-render, so it could only happen if no field was focused for e.g. half a second. 
- Add a "touched" property to all submission fields, that gets set to true when the user first interacts with a field
    - Ideally both on the backend and the UI (that way it would be saved across sessions), but only in the UI would still be a major improvement if it proves too much work
- Only show popup notifications after manually saving. Show a small unobtrusive message near the save button saying: "saved n seconds ago" that gets updated by both manual and auto-saves. If an auto save triggers an error or a warning, don't show it. Do update touched fields with the error style and message. If a manual save triggers an error or warning, use the same popup notifications we do now.
- Store values typed in the top row of a repeatable field on the client but not the server. Whenever the form is re-rendered, check whether there was something typed in the field before and reinstate it if it was. 
- Replace '+' on the button with the word 'Add', and ensure it is always right next to the input, to make it more clear that it will add the current value, rather than add another empty field below.
- All of this can be done in an estimated **40 hours**

**Comment by atarix83** [comment]
Our proposal to try to solve problems are : 

- add a configuration to eventually disable autosave on a timer, in the way institution can choose if use this functionality or not
- add a mechanism to detect conflicts between server state of the form and client state, and eventually re-render the form only if is necessary. Additional info can come from the server in case of automatic enrichment via extraction with identifiers or even for competitor edits. The rules that defined when a re-render is necessary are: 
1.  If the field has not been touched in the form after autosave wins the server state and form is re-rendered

1. If the field has been touched, you have saved the initial post-autosave state, if the server sends you back the same state the client wins and the form is not re-rendered

1. if the server reports a different status, it reports a conflict and form is re-rendered (by default it keeps what is on the client)

1. If a new autosave starts or the user ignores the conflicts, the client status will be maintained by adding only the additional information received from the server that was never touched by the client.

- an additional property could be added to the submission state, that keep track the focus on a field, so it should be easily restored it after 

> Only show popup notifications after manually saving. Show a small unobtrusive message near the save button saying: "saved n seconds ago" that gets updated by both manual and auto-saves. If an auto save triggers an error or a warning, don't show it. Do update touched fields with the error style and message. If a manual save triggers an error or warning, use the same popup notifications we do now.


we agree with this proposal

> Store values typed in the top row of a repeatable field on the client but not the server. Whenever the form is re-rendered, check whether there was something typed in the field before and reinstate it if it was.
Replace '+' on the button with the word 'Add', and ensure it is always right next to the input, to make it more clear that it will add the current value, rather than add another empty field below.



See  more detail on https://github.com/DSpace/dspace-angular/issues/864

Estimation for this is about 40 hours.

**Comment by tdonohue** [comment]
In today's meeting, we discussed the two proposals above.  We've decided to limit the scope of this ticket to **client side changes only**, and therefore are going with the approach suggested by @artlowel above (copied here for reference, with a few minor notes inline).  

> * Stop autosaving on a timer, instead auto-save whenever a field loses focus

We noted today that this change is potentially optional.  Once all other changes below are made, the auto-save on a timer would be _much less distruptive_ to the submitter.  Therefore, we could (as suggested by @atarix83) allow for the timer to remain, with an option to disable it.

> * Don't re-render the form when a server response comes in if any field is currently in focus, only when a field loses focus.
>     * this could still mean we lose the tabindex, meaning if you switch fields by using the tab button (something required for a lot accessibility use cases) you could lose your place when the autosave happens
>     * To counter that we could debounce the re-render, so it could only happen if no field was focused for e.g. half a second.
> * Add a "touched" property to all submission fields, that gets set to true when the user first interacts with a field
Ideally both on the backend and the UI (that way it would be saved across sessions), but only in the UI would still be a major improvement if it proves too much work

As noted in today's meeting, this change would be _limited to the client side_, as the primary goal of this "touched" property is simply to tell whether to put a red outline around a required field.  If that required field is not yet touched, then no red outline is needed.  If the required field was touched, but no value is entered, then a red outline should appear. 

> * Only show popup notifications after manually saving. Show a small unobtrusive message near the save button saying: "saved n seconds ago" that gets updated by both manual and auto-saves. If an auto save triggers an error or a warning, don't show it.
> * Do update touched fields with the error style and message. If a manual save triggers an error or warning, use the same popup notifications we do now.
> * Store values typed in the top row of a repeatable field on the client but not the server. Whenever the form is re-rendered, check whether there was something typed in the field before and reinstate it if it was.
> * Replace '+' on the button with the word 'Add', and ensure it is always right next to the input, to make it more clear that it will add the current value, rather than add another empty field below.

As noted in today's meeting, changing this field to an `Add` would exactly match the behavior of the DSpace 6 XMLUI.  In the DSpace 6 XMLUI, repeatable fields have an "Add" button and you must press it to save the value.  Obviously, in a future 7.x release we may be able to improve this default behavior, but it's "good enough" for now.

**The original estimate of 40 hours will be kept, as we've limited this ticket to UI level changes only.**

Tasks moved to other tickets:
* Because this ticket is limited to UI changes only, we may want to temporarily *disable* or rethink "automatic enrichment via extraction with identifiers".  _A separate ticket will be created and linked here when ready._
* @atarix83's proposal of comparing changes on server & UI may be worth revisiting after 7.0, as it might prove to be a better solution long term. However, we've delayed this work, as it's unclear (currently) which scenarios may require this sort of detailed check & how to resolve conflicts that arise. _A separate ticket will be created and linked here when ready._

**Comment by corrad82-4s** [comment]
@tdonohue we are available to work on this, for us is ok even if other community members carry on this issue, if a pull request will be ready within two weeks. 
Please let us know, we are ready to start development next Monday. 
Thanks

**Comment by tdonohue** [comment]
@corrad82-4s : That sounds fine. I've assigned this to your for your team to start work on & moved it to the "in progress" column.

NOTE for the 4Science team (@abollini , @atarix83 ): please be aware that I still consider #787 to be the **highest priority** of all tasks assigned to your team.  So, while you are welcome to work on this #835 ticket in parallel to the Angular 10 upgrade, please do not remove resources from the Angular 10 upgrade in order to work on this ticket.
